book/chapters/appendices/solutions.qmd

```diff
@@ -800,72 +801,174 @@ system.time(resample(task, learner, resampling))
+
+```{r}
+learner = lrn("classif.log_reg", predict_type = "prob") # otherwise our meausre does not work
+task = tsk("sonar")
+rr = resample(task, learner, rsmp("cv", folds = 5))
+```
+
+We now score the resample result using our custom measure and `msr("classif.acc")`.
+
+```{r}
+score = rr$score(list(msr_mae_prob, msr("classif.acc")))
+```
+
+In this case, there is a clear relationship between the classification accuracy and our custom measure, i.e. the higher the accuracy, the lower the mean absolute error of the predicted probabilities.
+
+```{r}
+cor(score$mae_prob, score$classif.acc)
+```
+
+
+3. "Tune" the `error_train` hyperparameter of the `classif.debug` learner on a continuous interval from 0 to 1, using a simple classification tree as the fallback learner and the penguins task.
+  Tune for 50 iterations using random search and 10-fold cross-validation.
+  Inspect the resulting archive and find out which evaluations resulted in an error, and which did not.
+  Now do the same in the interval 0.3 to 0.7.
+  Are your results surprising?
+
+First, we create the learner that we want to tune, mark the relevant parameter for tuning and set the fallback learner to a classification tree.
+
+```{r}
+lrn_debug = lrn("classif.debug",
+  error_train = to_tune(0, 1),
+  fallback = lrn("classif.rpart")
+)
+lrn_debug
+```
+
+This example is unusual, because we expect better results from the fallback classification tree than from the primary debug learner, which predicts the mode of the target distribution.
+Nonetheless it serves as a good example to illustrate the effects of training errors on the tuning results.
+
+We proceed with optimizing the classification accuracy of the learner on the penguings task.
+
+```{r}
+instance = tune(
+  learner =  lrn_debug,
+  task = tsk("penguins"),
+  resampling = rsmp("cv"),
+  tuner = tnr("random_search"),
+  measure = msr("classif.acc"),
+  term_evals = 50
+)
+instance
+```
+
+To find out which evaluations resulted in an error, we can inspect the `$archive` slot of the instance, which we convert to a `data.table` for easier filtering.
+
+```{r}
+archive = as.data.table(instance$archive)
+archive[, c("error_train", "classif.acc", "errors")]
+```
+
+Below, we visualize the relationship between the error probabilty and the classification accuracy.
+
+```{r}
+ggplot(data = archive, aes(x = error_train, y = classif.acc, color = errors)) +
+  geom_point() +
+  theme_minimal()
+```
+
+Higher values for `error_train` lead to more resampling iterations using the classification tree fallback learner and therefore to better classification accuracies.
+Therefore, the best found hyperparameter configurations will tend to have values of `error_train` close to 1.
+When multiple parameter configurations have the same test performance, the first one is chosen by `$result_learner_param_vals`.
+
+```{r}
+instance$result_learner_param_vals
+```
+
+We repeat the same experiment for the tuning interval from 0.3 to 0.7.
```

Comment:
Not sure what the point of this is -- results are essentially the same.

Reply:
I kind of agree, but it is like that in the book now.